analysis/src/vm/romgen.rs

```diff
@@ -0,0 +1,217 @@
+            batches
+                .first_mut()
+                .expect("function should have at least one statement as it must return")
+                .statements
+                .insert(
+                    0,
+                    parse_function_statement(&format!("_{}::", function.name)),
+                );
+
+            // modify the last batch to be caused by the coming label
+            let last = batches
+                .last_mut()
+                .expect("function should have at least one statement as it must return");
+            last.set_reason(IncompatibleSet::from(Incompatible::Label));
+
+            rom.extend(batches);
+        }
+
+        // THE FOLLOWING IS NECESSARY BECAUSE OF WITGEN, IT COULD BE REMOVED ONCE WITGEN CAN CALL THE INFINITE LOOP ITSELF
+        // we get the location of the sink so that witgen jumps to it after the first call is done
+        // this constrains the VM to being able to execute only one call, which will be fixed in the future
+        let sink_id = T::from(rom.len() as u64);
+        let latch = machine.latch.as_ref().unwrap();
+        let sigma = "_sigma";
+
+        machine.constraints.push(PilBlock {
+            start: 0,
+            statements: vec![
+                // declare `_sigma` as the sum of the latch, will be 0 and then 1 after the end of the first call
+                parse_pil_statement(&format!("col witness {sigma}")),
+                parse_pil_statement(&format!(
+                    "{sigma}' = (1 - first_step') * ({sigma} + {latch})"
+                )), // HACKY: THIS ASSUMES `first_step` is defined here!!
+                // once `_sigma` is 1, constrain `_function_id` to the label of the sink
+                parse_pil_statement(&format!("{sigma} * ({function_id} - {sink_id}) = 0")),
+            ],
+        });
+        ///////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+        rom.extend(vec![Batch::from(vec![
+            parse_function_statement("_sink::"),
+            parse_function_statement("_loop;"),
+        ])]);
+
+        machine.function_id = Some(function_id.into());
+
+        machine.rom = Some(Rom {
+            statements: rom.into_iter().collect(),
+        });
+
+        machine
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use number::Bn254Field;
+    use pretty_assertions::assert_eq;
+
+    use crate::vm::test_utils::generate_rom_str;
+
+    use super::*;
+
+    #[test]
+    fn vm() {
+        let vm = r#"
+            machine VM {
+
+                reg pc[@pc];
+                reg X[<=];
+                reg Y[<=];
+                reg Z[<=];
+                reg A;
+                reg B;
+            
+                instr add X, Y -> Z { X + Y = Z }
+            
+                instr sub X, Y -> Z { X - Y = Z }
+            
+                function add x: field, y: field -> field {
```

Comment:
I don't think function and instr names should be allowed to collide, it can look confusing

Reply:
Fair enough. Changed the source here. Added #489